pkg/utils/volumeattachment/volumeattachment.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volumeattachment
+
+import (
+	"context"
+
+	"github.com/samber/lo"
+
+	v1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	nodeutil "sigs.k8s.io/karpenter/pkg/utils/node"
+	"sigs.k8s.io/karpenter/pkg/utils/pod"
+	volumeutil "sigs.k8s.io/karpenter/pkg/utils/volume"
+)
+
+// FilterVolumeAttachments filters out volumeAttachments that should not block the termination of the passed node
+func FilterVolumeAttachments(ctx context.Context, kubeClient client.Client, node *v1.Node, volumeAttachments []*storagev1.VolumeAttachment, clk clock.Clock) ([]*storagev1.VolumeAttachment, error) {
+	var filteredVolumeAttachments []*storagev1.VolumeAttachment
+	// No need to filter empty volumeAttachments list
+	if len(volumeAttachments) == 0 {
+		return volumeAttachments, nil
+	}
+	// Filter out non-drain-able pods
+	pods, err := nodeutil.GetPods(ctx, kubeClient, node)
+	if err != nil {
+		return nil, err
+	}
+	drainablePods := lo.Filter(pods, func(p *v1.Pod, _ int) bool {
+		return pod.IsDrainable(p, clk)
+	})
+	// Filter out Multi-Attach volumes
+	shouldNotFilterOutVolume := make(map[string]bool)
+	for _, p := range drainablePods {
+		for _, v := range p.Spec.Volumes {
+			pvc, err := volumeutil.GetPersistentVolumeClaim(ctx, kubeClient, p, v)
+			if err != nil {
+				return nil, err
+			}
+			if pvc != nil {
+				shouldNotFilterOutVolume[pvc.Spec.VolumeName] = CannotMultiAttach(*pvc)
+			}
+		}
+	}
+	for i := range volumeAttachments {
+		pvName := volumeAttachments[i].Spec.Source.PersistentVolumeName
+		if pvName != nil && shouldNotFilterOutVolume[*pvName] {
+			filteredVolumeAttachments = append(filteredVolumeAttachments, volumeAttachments[i])
+		}
+	}
+	return filteredVolumeAttachments, nil
+}
+
+// CannotMultiAttach returns true if the persistentVolumeClaim's underlying volume cannot be attached to multiple nodes
+// i.e. its access mode is not ReadWriteOnce/ReadWriteOncePod
```

Comment:
TLDR: Deleting this code because the faster node termination is not worth orphaned volume attachments. This implicitly resolves this comment

---

Context: Stateful workloads necessitating multi-attach volumes DO NOT run into 6+ minute delays. This is because the rescheduled workload never gets the `FailedAttachVolume, Multi-attach error" because that volume IS allowed to be attached to multiple nodes at once (and therefore have more than 1 volume attachment object in kubernetes at a time) 

```
FailedAttachVolume  108s  attachdetach-controller  Multi-Attach error for volume "pvc-01998971-b963-4ea8-9170-53cb49837082" Volume is already exclusively attached to one node and can't be attached to another
```

The point of filtering out volumes that are not ReadWriteOnce/ReadWriteOncePod access mode (and therefore are node multi-attach) was so that we can terminate node quicker. 

However, the more I think about it, it's probably best to err on the side of waiting for volume attachment object to be cleaned up before deleting node object because we don't want orphaned volume attachment objects at all. Especially when Kubernetes v1.32 might NEVER force detach volume attachments. 

We can always add this optimization later.